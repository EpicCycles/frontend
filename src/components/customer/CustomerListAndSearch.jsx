import React, { Fragment } from 'react';
import * as PropTypes from 'prop-types';
import CustomerSearch from './CustomerSearch';
import CustomerListGridHeaders from './CustomerListGridHeaders';
import CustomerListGridRow from './CustomerListGridRow';
import Pagination from '../../common/pagination';
import AddLink from '../app/model/AddLink';
import { isItAnObject } from '../../helpers/utils';

const CustomerListAndSearch = props => {
  const {
    addNewCustomer,
    getCustomerList,
    getCustomerListPage,
    getCustomer,
    isLoading,
    customers,
    count,
    next,
    previous,
    searchParams,
  } = props;
  if (!isLoading) {
    if (!customers) {
      getCustomerList('', '', '');
    }
  }
  return (
    <Fragment>
      <h2>Find Customer</h2>
      <CustomerSearch
        getCustomerList={getCustomerList}
        searchParams={searchParams}
        isLoading={isLoading}
      />
      {count > 0 && (
        <div className="grid" key="customer-list-container">
          <CustomerListGridHeaders includeActions={true} />
          {customers.map(customer => (
            <CustomerListGridRow
              key={`cust-${customer.id}`}
              customer={customer}
              getCustomer={getCustomer}
            />
          ))}
        </div>
      )}
      <div className="row align_left">
        {count > 0 ? (
          <Pagination
            id="customer-pagination"
            previous={previous}
            next={next}
            count={count}
            getPage={getCustomerListPage}
          />
        ) : (
          <div>No Customer to show, set new criteria and search, or </div>
        )}
        <AddLink addFunction={addNewCustomer} addObjectName={'Customer'} />
      </div>
    </Fragment>
  );
};
CustomerListAndSearch.defaultProps = {
  count: 0,
};
CustomerListAndSearch.propTypes = {
  addNewCustomer: PropTypes.func.isRequired,
  getCustomerList: PropTypes.func.isRequired,
  getCustomerListPage: PropTypes.func.isRequired,
  getCustomer: PropTypes.func.isRequired,
  searchParams: PropTypes.object,
  isLoading: PropTypes.bool,
  customers: PropTypes.array,
  count: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  next: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  previous: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};
export default CustomerListAndSearch;
