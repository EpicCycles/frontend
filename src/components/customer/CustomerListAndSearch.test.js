import React from 'react';
import toJson from 'enzyme-to-json';
import CustomerListAndSearch from './CustomerListAndSearch';

describe('CustomerListAndSearch.index tests', () => {
  // our mock login function to replace the one provided by mapDispatchToProps

  const initialState = {
    customer: {
      count: 0,
      isLoading: false,
      customers: [],
      page: 1,
      perPage: 5,
    },
  };

  const customers = [
    {
      id: 16,
      first_name: 'Anna',
      last_name: 'Weaver',
      email: 'anna.weaver@johnlewis.co.uk',
      add_date: '2018-07-04T13:02:09.988286+01:00',
      upd_date: '2018-07-04T13:02:09.988343+01:00',
    },
    {
      id: 26,
      first_name: 'dfds',
      last_name: 'Weasdfdsfver',
      add_date: '2018-07-04T13:02:09.988286+01:00',
      upd_date: '2018-07-04T13:02:09.988343+01:00',
    },
  ];
  it('renders the CustomerListAndSearch correctly', () => {
    const mockGetCustomerListPagefn = jest.fn();
    const mockGetCustomerListfn = jest.fn();
    const initialList = shallow(
      <CustomerListAndSearch
        getCustomerListPage={mockGetCustomerListPagefn}
        getCustomerList={mockGetCustomerListfn}
        clearCustomerState={jest.fn()}
        addNewCustomer={jest.fn()}
        getCustomer={jest.fn()}
        count={0}
        f
        isLoading={false}
        page={0}
        totalPages={0}
      />,
    );
    expect(toJson(initialList)).toMatchSnapshot();
    expect(mockGetCustomerListfn).toHaveBeenCalledWith('', '', '');
    expect(mockGetCustomerListPagefn).not.toHaveBeenCalled();
  });
  it('renders the CustomerListAndSearch correctly when there are customers', () => {
    const mockGetCustomerListPagefn = jest.fn();
    const mockGetCustomerListfn = jest.fn();
    const initialList = shallow(
      <CustomerListAndSearch
        getCustomerListPage={mockGetCustomerListPagefn}
        getCustomerList={mockGetCustomerListfn}
        addNewCustomer={jest.fn()}
        clearCustomerState={jest.fn()}
        getCustomer={jest.fn()}
        count={6}
        customers={customers}
        isLoading={false}
        page={0}
        totalPages={1}
        searchParams={{ firstName: 'fred' }}
      />,
    );
    expect(toJson(initialList)).toMatchSnapshot();
    expect(mockGetCustomerListfn).not.toHaveBeenCalled();
    expect(mockGetCustomerListPagefn).not.toHaveBeenCalled();
  });
});
