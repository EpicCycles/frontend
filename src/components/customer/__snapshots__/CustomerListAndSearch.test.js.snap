// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerListAndSearch.index tests renders the CustomerListAndSearch correctly 1`] = `
<Fragment>
  <h2>
    Find Customer
  </h2>
  <CustomerSearch
    className="row"
    isLoading={false}
    searchParams={
      Object {
        "email": "",
        "firstName": "",
        "lastName": "",
      }
    }
  />
  <div
    className="row align_left"
  >
    <div>
      No Customer to show, set new criteria and search, or 
    </div>
    <AddLink
      addFunction={[MockFunction]}
      addObjectName="Customer"
    />
  </div>
</Fragment>
`;

exports[`CustomerListAndSearch.index tests renders the CustomerListAndSearch correctly when there are customers 1`] = `
<Fragment>
  <h2>
    Find Customer
  </h2>
  <CustomerSearch
    className="row"
    isLoading={false}
    searchParams={
      Object {
        "email": "",
        "firstName": "",
        "lastName": "",
      }
    }
  />
  <div
    className="grid"
    key="customer-list-container"
  >
    <CustomerListGridHeaders
      className=""
      includeActions={true}
    />
    <CustomerListGridRow
      customer={
        Object {
          "add_date": "2018-07-04T13:02:09.988286+01:00",
          "email": "anna.weaver@johnlewis.co.uk",
          "first_name": "Anna",
          "id": 16,
          "last_name": "Weaver",
          "upd_date": "2018-07-04T13:02:09.988343+01:00",
        }
      }
      getCustomer={[MockFunction]}
      key="cust-16"
    />
    <CustomerListGridRow
      customer={
        Object {
          "add_date": "2018-07-04T13:02:09.988286+01:00",
          "first_name": "dfds",
          "id": 26,
          "last_name": "Weasdfdsfver",
          "upd_date": "2018-07-04T13:02:09.988343+01:00",
        }
      }
      getCustomer={[MockFunction]}
      key="cust-26"
    />
  </div>
  <div
    className="row align_left"
  >
    <Pagination
      count={6}
      id="customer-pagination"
    />
    <AddLink
      addFunction={[MockFunction]}
      addObjectName="Customer"
    />
  </div>
</Fragment>
`;
